Math/NumberTheory/ArithmeticFunctions/Inverse.hs

```diff
@@ -0,0 +1,354 @@
+-- >>> inverseTotient (const $ Const 1) 120 :: Const (Sum Word) Word
+-- Const (Sum {getSum = 17})
+--
+-- Find minimal and maximal preimages:
+--
+-- >>> inverseTotient MinWord 120 :: MinWord
+-- MinWord {unMinWord = 143}
+-- >>> inverseTotient MaxWord 120 :: MaxWord
+-- MaxWord {unMaxWord = 462}
+inverseTotient
+  :: (Semiring b, Euclidean a, UniqueFactorisation a, Ord a)
+  => (a -> b)
+  -> a
+  -> b
+inverseTotient point n = invertFunction point totientA invTotient n
+{-# SPECIALISE inverseTotient :: Semiring b => (Int -> b) -> Int -> b #-}
+{-# SPECIALISE inverseTotient :: Semiring b => (Word -> b) -> Word -> b #-}
+{-# SPECIALISE inverseTotient :: Semiring b => (Integer -> b) -> Integer -> b #-}
+{-# SPECIALISE inverseTotient :: Semiring b => (Natural -> b) -> Natural -> b #-}
+
+-- | The inverse 'sigma' 1 function such that
+--
+-- > all ((== x) . sigma 1) (inverseSigma x)
+-- > x `elem` inverseSigma (sigma 1 x)
+--
+-- The return value is parametrized by a semiring, which allows
+-- various applications. E. g., list all preimages:
+--
+-- >>> import qualified Data.Set as S
+-- >>> import Data.Semigroup
+-- >>> S.map getProduct (inverseSigma (S.singleton . Product) 120) :: S.Set Word
+-- fromList [54,56,87,95]
+--
+-- Count preimages:
+--
+-- >>> import Control.Applicative
+-- >>> import Data.Semigroup
+-- >>> inverseSigma (const $ Const 1) 120 :: Const (Sum Word) Word
+-- Const (Sum {getSum = 4})
+--
+-- Find minimal and maximal preimages:
+--
+-- >>> inverseSigma MinWord 120 :: MinWord
+-- MinWord {unMinWord = 54}
+-- >>> inverseSigma MaxWord 120 :: MaxWord
+-- MaxWord {unMaxWord = 95}
+inverseSigma
+  :: (Semiring b, Euclidean a, UniqueFactorisation a, Integral a)
+  => (a -> b)
+  -> a
+  -> b
+inverseSigma point n = invertFunction point (sigmaA 1) invSigma n
+{-# SPECIALISE inverseSigma :: Semiring b => (Int -> b) -> Int -> b #-}
+{-# SPECIALISE inverseSigma :: Semiring b => (Word -> b) -> Word -> b #-}
+{-# SPECIALISE inverseSigma :: Semiring b => (Integer -> b) -> Integer -> b #-}
+{-# SPECIALISE inverseSigma :: Semiring b => (Natural -> b) -> Natural -> b #-}
+
+--------------------------------------------------------------------------------
+-- Wrappers
+
+-- | Wrapper to use in conjunction with 'inverseTotient' and 'inverseSigma'.
+-- Extracts the maximal preimage of function.
+newtype MaxWord = MaxWord { unMaxWord :: Word }
+  deriving (Show)
+
+instance Semiring MaxWord where
+  zero = MaxWord minBound
+  one  = MaxWord 1
+  plus  (MaxWord a) (MaxWord b) = MaxWord (a `max` b)
+  times (MaxWord a) (MaxWord b) = MaxWord (a * b)
+
+-- | Wrapper to use in conjunction with 'inverseTotient' and 'inverseSigma'.
+-- Extracts the minimal preimage of function.
+newtype MinWord = MinWord { unMinWord :: Word }
+  deriving (Show)
+
+instance Semiring MinWord where
+  zero = MinWord maxBound
+  one  = MinWord 1
+  plus  (MinWord a) (MinWord b) = MinWord (a `min` b)
+  times (MinWord a) (MinWord b) = MinWord (a * b)
```

Comment:
TIL this how we define one valid semiring for the `min` operator, I wasn't quite sure `(*)` would work.